tests/tests_pytorch/utilities/test_imports.py

```diff
@@ -67,3 +76,62 @@ def test_imports():
         assert not _POPTORCH_AVAILABLE
     else:
         assert _POPTORCH_AVAILABLE
+
+
+def shortcut_patch(orig_fn, shortcut_case, attr_names=None):
+    def new_fn(*args, **kwargs):
+        if attr_names is not None:
+            self = args[0]
+            values = tuple(getattr(self, attr_name) for attr_name in attr_names)
+        else:
+            values = args
+        match = True
+        for value, case in zip(values, shortcut_case):
+            if value != case:
+                match = False
+                break
+        if match:
+            return False
+        return orig_fn(*args, **kwargs)
+
+    return new_fn
+
+
+@pytest.mark.parametrize(
+    ["patch_name", "new_fn", "to_import"],
+    [
+        ("torch.distributed.is_available", shortcut_patch(is_available, ()), "pytorch_lightning"),
+        (
+            "lightning_utilities.core.imports.RequirementCache.__bool__",
+            shortcut_patch(RequirementCache.__bool__, ("neptune-client",), ("requirement",)),
+            "pytorch_lightning.loggers.neptune",
+        ),
+        (
+            "lightning_utilities.core.imports.RequirementCache.__bool__",
+            shortcut_patch(RequirementCache.__bool__, ("jsonargparse[signatures]>=4.12.0",), ("requirement",)),
+            "pytorch_lightning.cli",
+        ),
+        (
+            "lightning_utilities.core.imports.module_available",
+            shortcut_patch(module_available, ("fairscale.nn",)),
+            "pytorch_lightning.strategies",
+        ),
+        (
+            "lightning_utilities.core.imports.compare_version",
+            shortcut_patch(compare_version, ("torch", operator.ge, "1.12.0")),
+            "pytorch_lightning.strategies.fully_sharded_native",
+        ),
+    ],
+    ids=["ProcessGroup", "neptune", "cli", "fairscale", "fully_sharded_native"],
+)
+def test_import_with_unavailable_dependencies(patch_name, new_fn, to_import):
+    pl_keys = list(
+        key for key in sys.modules.keys() if key.startswith("pytorch_lightning") or key.startswith("lightning")
+    )
+    for pl_key in pl_keys:
+        try:
+            del sys.modules[pl_key]
```

Comment:
```suggestion
            del sys.modules.pop(pl_key, None)
```

currently this marks it to be deleted for the gc, doesn't necessarily mean it will be deleted directly. 
With popping this ensures that the key is definitely not present in sys.modules anymore. Also this way you don't need the exception anymore :)

To be sure it has been deleted, we might want to add a `gc.collect()` after the loop.

Reply:
We don't actually need it to be deleted in the sense of `gc` and python interpreter, we need it to *not* be present in `sys.modules` - if the module is not present, `importlib` machinery will reload it as it doesn't have a place from which it would find a cached module.

Reply:
But, I agree, popping is nicer since it's not necessary to use `try...catch` then :)